src/Workspaces/Remote/Core/BrokeredServiceConnection.cs

```diff
@@ -293,22 +293,33 @@ public override async ValueTask<Optional<TResult>> TryInvokeAsync<TResult>(Solut
             //  6. 'pipe.Writer' is completed
             //  7. 'pipe.Reader' is completed
             //  8. OperationCanceledException is thrown back to the caller
+            //
+            // Since StreamJsonRpc will automatically complete writers when an error occurs (and cancellation is
```

Comment:
> will automatically complete writers

It isn't "writers" that we complete. It's whatever you pass in for StreamJsonRpc to own. Whether you pass in `PipeReader` or `PipeWriter` or `IDuplexPipe` as an argument, you expect the receiver to access it and call Complete on them when done. So yes, in error conditions (including cancellation) JsonRpc will do this by disposing the underlying `MultiplexingStream.Channel`, assuming I suppose that we got far enough along to send the request in the first place. 
In _non_-error conditions, JsonRpc will also do this (or rather `MultiplexingStream` will) by relaying completion done on the receiver side. 

In success cases, the *caller* should not complete the reader/writer/pipe that they pass in as arguments since they are yielding these objects' ownership to another party. Completing these objects when you don't own them is generally hazardous because these objects are not thread-safe, so if you call Complete on them while another thread is using them or their buffers, you could end up corrupting data in arrays that were recycled and have since been assigned to another owner.
In failure cases, JsonRpc would normally have completed them as well, but that's not a guarantee I suppose because InvokeAsync may have thrown an exception before ever getting to the point of processing those pipe arguments.

Reply:
➡️ Now rewritten